[ExecuTorch][Export][1/N] Export API pipeline re-architecture, making it composable

abhinaykukkadapu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #12938
* #12937
* __->__ #12936

RFC: https://github.com/pytorch/executorch/issues/12660

This diff introduces composable architecture for the export pipeline.

Changes:
1. Introduces notion of `PipelineArtifact` which encompasses stage artifacts and run context necessary to execute a stage and the pipeline.
1. Move all stages to new file to better architect the component
1. Export api will now accept `pipeline_stages` and honors the sequence if provided otherwise fallback to default sequence: `source_transform -> quantize -> torch.export -> to_edge_transform_and_lower -> to_executorch`
1. Validate if pipeline provided is valid, error out if not.
1. Add unittests to test pipeline sequence and stages

With this one can execute a partial pipeline, for example, one can just do `TORCH_EXPORT -> TO_EDGE_TRANSFORM_AND_LOWER -> TO_EXECUTORCH`

Current limitation:
1. `TORCH_EXPORT` stage is mandatory to avoid someone passing incorrect input such as `ExportedProgram` instead of nn.module, this enforcement will anchor the pipeline to have expected stages to run.
   - This limitation will be removed if we choose to add the support to take `ExportedProgram` as input.

Note:
1. this diff is large because I moved stages to separate file and added bunch of unittests but core functionality that is added can be reviewed in `export.py` file. CC: @digantdesai
2. Export component was supposed to mimic stages design in tester and now it is.

Fixes: #12928

Differential Revision: [D79120574](https://our.internmc.facebook.com/intern/diff/D79120574/)

abhinaykukkadapu committed:
[ExecuTorch][Export][1/N] Export API pipeline re-architecture, making it composable

RFC: https://github.com/pytorch/executorch/issues/12660

This diff introduces composable architecture for the export pipeline.

Changes:
1. Introduces notion of `PipelineArtifact` which encompasses stage artifacts and run context necessary to execute a stage and the pipeline.
1. Move all stages to new file to better architect the component
1. Export api will now accept `pipeline_stages` and honors the sequence if provided otherwise fallback to default sequence: `source_transform -> quantize -> torch.export -> to_edge_transform_and_lower -> to_executorch`
1. Validate if pipeline provided is valid, error out if not.
1. Add unittests to test pipeline sequence and stages

With this one can execute a partial pipeline, for example, one can just do `TORCH_EXPORT -> TO_EDGE_TRANSFORM_AND_LOWER -> TO_EXECUTORCH`

Current limitation:
1. `TORCH_EXPORT` stage is mandatory to avoid someone passing incorrect input such as `ExportedProgram` instead of nn.module, this enforcement will anchor the pipeline to have expected stages to run.
   - This limitation will be removed if we choose to add the support to take `ExportedProgram` as input.

Note:
1. this diff is large because I moved stages to separate file and added bunch of unittests but core functionality that is added can be reviewed in `export.py` file. CC: @digantdesai
2. Export component was supposed to mimic stages design in tester and now it is.

Fixes: #12928

Differential Revision: [D79120574](https://our.internmc.facebook.com/intern/diff/D79120574/)

[ghstack-poisoned]

abhinaykukkadapu committed:
Update on "[ExecuTorch][Export][1/N] Export API pipeline re-architecture, making it composable"


RFC: https://github.com/pytorch/executorch/issues/12660

This diff introduces composable architecture for the export pipeline.

Changes:
1. Introduces notion of `PipelineArtifact` which encompasses stage artifacts and run context necessary to execute a stage and the pipeline.
1. Move all stages to new file to better architect the component
1. Export api will now accept `pipeline_stages` and honors the sequence if provided otherwise fallback to default sequence: `source_transform -> quantize -> torch.export -> to_edge_transform_and_lower -> to_executorch`
1. Validate if pipeline provided is valid, error out if not.
1. Add unittests to test pipeline sequence and stages

With this one can execute a partial pipeline, for example, one can just do `TORCH_EXPORT -> TO_EDGE_TRANSFORM_AND_LOWER -> TO_EXECUTORCH`

Current limitation:
1. `TORCH_EXPORT` stage is mandatory to avoid someone passing incorrect input such as `ExportedProgram` instead of nn.module, this enforcement will anchor the pipeline to have expected stages to run.
   - This limitation will be removed if we choose to add the support to take `ExportedProgram` as input.

Note:
1. this diff is large because I moved stages to separate file and added bunch of unittests but core functionality that is added can be reviewed in `export.py` file. CC: digantdesai
2. Export component was supposed to mimic stages design in tester and now it is.

Fixes: #12928

Differential Revision: [D79120574](https://our.internmc.facebook.com/intern/diff/D79120574/)

[ghstack-poisoned]

digantdesai reviewed (approved):
LGTM. Thanks for refactoring. No functional changes right?

larryliu0820 reviewed (approved):
Reviewed internally

abhinaykukkadapu committed:
Update on "[ExecuTorch][Export][1/N] Export API pipeline re-architecture, making it composable"


RFC: https://github.com/pytorch/executorch/issues/12660

This diff introduces composable architecture for the export pipeline.

Changes:
1. Introduces notion of `PipelineArtifact` which encompasses stage artifacts and run context necessary to execute a stage and the pipeline.
1. Move all stages to new file to better architect the component
1. Export api will now accept `pipeline_stages` and honors the sequence if provided otherwise fallback to default sequence: `source_transform -> quantize -> torch.export -> to_edge_transform_and_lower -> to_executorch`
1. Validate if pipeline provided is valid, error out if not.
1. Add unittests to test pipeline sequence and stages

With this one can execute a partial pipeline, for example, one can just do `TORCH_EXPORT -> TO_EDGE_TRANSFORM_AND_LOWER -> TO_EXECUTORCH`

Current limitation:
1. `TORCH_EXPORT` stage is mandatory to avoid someone passing incorrect input such as `ExportedProgram` instead of nn.module, this enforcement will anchor the pipeline to have expected stages to run.
   - This limitation will be removed if we choose to add the support to take `ExportedProgram` as input.

Note:
1. this diff is large because I moved stages to separate file and added bunch of unittests but core functionality that is added can be reviewed in `export.py` file. CC: digantdesai
2. Export component was supposed to mimic stages design in tester and now it is.

Fixes: #12928

Differential Revision: [D79120574](https://our.internmc.facebook.com/intern/diff/D79120574/)

[ghstack-poisoned]

abhinaykukkadapu committed:
Update on "[ExecuTorch][Export][1/N] Export API pipeline re-architecture, making it composable"


RFC: https://github.com/pytorch/executorch/issues/12660

This diff introduces composable architecture for the export pipeline.

Changes:
1. Introduces notion of `PipelineArtifact` which encompasses stage artifacts and run context necessary to execute a stage and the pipeline.
1. Move all stages to new file to better architect the component
1. Export api will now accept `pipeline_stages` and honors the sequence if provided otherwise fallback to default sequence: `source_transform -> quantize -> torch.export -> to_edge_transform_and_lower -> to_executorch`
1. Validate if pipeline provided is valid, error out if not.
1. Add unittests to test pipeline sequence and stages

With this one can execute a partial pipeline, for example, one can just do `TORCH_EXPORT -> TO_EDGE_TRANSFORM_AND_LOWER -> TO_EXECUTORCH`

Current limitation:
1. `TORCH_EXPORT` stage is mandatory to avoid someone passing incorrect input such as `ExportedProgram` instead of nn.module, this enforcement will anchor the pipeline to have expected stages to run.
   - This limitation will be removed if we choose to add the support to take `ExportedProgram` as input.

Note:
1. this diff is large because I moved stages to separate file and added bunch of unittests but core functionality that is added can be reviewed in `export.py` file. CC: digantdesai
2. Export component was supposed to mimic stages design in tester and now it is.

Fixes: #12928

Differential Revision: [D79120574](https://our.internmc.facebook.com/intern/diff/D79120574/)

[ghstack-poisoned]
